Support updating of MP OpenAPI Document and UI Endpoint

abutch3r opened the pull request:
Provide a common class that can handle both MP OpenAPI endpoints.

Endpoints and validated to ensure meet restricted requirements

Add a mpOpenAPI Server configuration element that can take values for either endpoint.

Update backend bundles to allow for custom endpoint allocation. Both for MP OpenAPI 1.0-1.1 (com.ibm.ws.microprofile.openapi) and MP OpenAPI 2.0-3.1+ (io.openliberty.microprofile.openapi.2.0.internal)

Add CustomUITest to Test UI Path updates

Enablement is guarded both via Beta check and additional system property that must be set to follow none-default behavior

Fixes #25072
Fixes #25073
Fixes #25074
Fixes #25075 
Fixes #25375
Fixes #25076
Some Fixes for #25437 
Some Fixes for #25439 
Includes the changes from #25511

abutch3r committed:
Support updating of MP OpenAPI Document and UI Endpoint

Provide a common class that can handle both Doc and UI endpoints and allows for validation of both endpoints and handles failure cases.

Add a mpOpenAPI Server configuration element that can take values for either endpoint. With support for live modification by unregistering the created WABs and creating new ones if the config changes

Update backend bundles to allow for custom endpoint allocation. Both for MP OpenAPI 1.0-1.1 (com.ibm.ws.microprofile.openapi) and MP OpenAPI 2.0-3.1+ (io.openliberty.microprofile.openapi.2.0.internal)

Add CustomUITest to Test UI Path updates

Add new messages to handle range of useful information and failure cases

Improve URL checking for OpenAPI document endpoint with removal of hardcoded /openapi reference

abutch3r commented:
Messages are taken from #25261 with minor updates to support both doc and UI messages.

abutch3r commented:
#libby #build

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/bnd.bnd:
Mix of tabs and spaces for indentation in this file. I don't mind which is used.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,27 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
 OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W.explanation=The server is configured to create OpenAPI documentation for the application or module name, but no application or web module with that name was deployed.
 OPENAPI_MERGE_UNUSED_INCLUDE_CWWKO1667W.useraction=Check that all applications started correctly and that the names in the configuration property are specified correctly.
+
+OPEN_API_PATH_UPDATE_CWWKO1668I=CWWKO1668I: The OpenAPI {0} path has been set to {1}.
+OPEN_API_PATH_UPDATE_CWWKO1668I.explanation=The OpenAPI UI/Document path has been set to a value that is not the default.
+OPEN_API_PATH_UPDATE_CWWKO1668I.useraction=No action is required.
```
I think this needs to be two messages for translation purposes.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,172 @@
+                resolvePathFromProperties(properties);
+            }
+        }
+        UIWabConfigManager = new WABConfigManager(context, OPEN_API_UI_VAR_NAME, uiPath, "OpenAPI UI");
+        DocWabConfigManager = new WABConfigManager(context, OPEN_API_DOC_VAR_NAME, docPath, "OpenAPI Doc");
+        UIWabConfigManager.register();
+        DocWabConfigManager.register();
```
If this is called in response to a config modification, we need to unregister the old config managers (or update them rather than creating new ones).

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,172 @@
+    private String uiPath = null;
+    private String docPath = null;
+
+    private WABConfigManager UIWabConfigManager;
+    private WABConfigManager DocWabConfigManager;
```
Field names should not start with a capital letter.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,172 @@
+            if (uiPath.equals(docPath)) {
+                Tr.error(tc, "OPEN_API_PATH_CONFLICT_CWWKO1670E", uiPath);
+                uiPath=docPath+OPEN_API_UI_PATH;
+            } else if (!validatePath(uiPath, "UI")){
+                    Tr.error(tc,"OPEN_API_PATH_UPDATE_FAILED_CWWKO1669E", "UI");
+                    uiPath=docPath+OPEN_API_UI_PATH;
+            }
```
Indentation is off, looks like the file needs formatted.

abutch3r commented:
#libby #build

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.explanation=The OpenAPI Document path has been set to a value that is not the default.
+OPEN_API_DOC_PATH_UPDATE_CWWKO1669I.useraction=No action is required.
+
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E=CWWKO1670E: The OpenAPI UI path is invalid, the default value will be used instead. Check previous warnings for cause of validation failure.
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation= The OpenAPI UI path is invalid and will use the default value instead.
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.useraction=Check earlier error messages for causes of validation failures.
```
Modify punctuation and word choice for ease of reading.

L2 suggested edits:
```
OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E=CWWKO1670E: The OpenAPI UI path is invalid. The default value will be used instead. Check previous warnings for the cause of validation failure.
OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation= The OpenAPI UI path is invalid.  The default value will be used instead.
OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.useraction=Check earlier error messages for the cause of validation failure.
```

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.explanation= The OpenAPI UI path is invalid and will use the default value instead.
+OPEN_API_UI_PATH_UPDATE_FAILED_CWWKO1670E.useraction=Check earlier error messages for causes of validation failures.
+
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E=CWWKO1671E: The OpenAPI Document path is invalid, the default value will be used instead. Check previous warnings for cause of validation failure.
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.explanation= The OpenAPI Document path is invalid and will use the default value instead.
+OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.useraction=Check earlier error messages for causes of validation failures.
```
Modify punctuation and word choice for ease of reading.

L2 suggested edits:
```
OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E=CWWKO1671E: The OpenAPI Document path is invalid.  The default value will be used instead. Check previous warnings for the cause of validation failure.
OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.explanation= The OpenAPI Document path is invalid.  The default value will be used instead.
OPEN_API_DOC_PATH_UPDATE_FAILED_CWWKO1671E.useraction=Check earlier error messages for the cause of validation failure.
```

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_CONFLICT_CWWKO1672E.explanation=The OpenAPI Document and OpenAPI UI paths cannot be the same.
+OPEN_API_UI_PATH_CONFLICT_CWWKO1672E.useraction=Specify a unique path for the OpenAPI UI.
+
+OPEN_API_UI_PATH_POTENTIAL_CONFLICT_CWWKO1673W=CWWKO1673W: The OpenAPI UI path was set to the same path as the {0} feature, which might prevent startup of itself or the other feature.
```
L2 suggestion to modify word choice for ease of reading:
`OPEN_API_UI_PATH_POTENTIAL_CONFLICT_CWWKO1673W=CWWKO1673W: The OpenAPI UI path was set to the same path as the {0} feature, which might prevent the startup of itself or other features.`

helyarp replied:
ID review:  Either "prevent startup" or "prevent the startup" is acceptable. I prefer "prevent startup"

Azquelt replied:
Leaving this as-is

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_CONFLICT_CWWKO1672E.useraction=Specify a unique path for the OpenAPI UI.
+
+OPEN_API_UI_PATH_POTENTIAL_CONFLICT_CWWKO1673W=CWWKO1673W: The OpenAPI UI path was set to the same path as the {0} feature, which might prevent startup of itself or the other feature.
+OPEN_API_UI_PATH_POTENTIAL_CONFLICT_CWWKO1673W.explanation=An OpenAPI UI path was set to the same path as another feature, which can prevent startup or access.
```
Can "access" here be more specific? What kind of access?

Azquelt replied:
Both OpenAPI and another feature are trying to serve some data at the same URL path. One or other of them won't be accessible to users.

I'll try to rewrite this to be more clear.

Azquelt replied:
I'm going to remove this message for now. We'll revisit it in #25367.

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_DOC_PATH_POTENTIAL_CONFLICT_CWWKO1674W.explanation=An OpenAPI Document path was set to the same path as another feature, which can prevent startup or access.
+OPEN_API_DOC_PATH_POTENTIAL_CONFLICT_CWWKO1674W.useraction=Specify a different path for the OpenAPI path.
+
+OPEN_API_UI_PATH_INVALID_CWWKO1675E=CWWKO1675E: The OpenAPI UI path is invalid or contains invalid characters. The URL must be constructed using Unicode alphanumeric characters A-Za-z0-9, the underscore (_), dash (-), forward slash (/) and period (.).
```
L2 suggested edit to remove the word "the" prior to "underscore", to maintain parallelism in the list of valid characters:
`OPEN_API_UI_PATH_INVALID_CWWKO1675E=CWWKO1675E: The OpenAPI UI path is invalid or contains invalid characters. The URL must be constructed using Unicode alphanumeric characters A-Za-z0-9, underscore (_), dash (-), forward slash (/), or period (.).`

helyarp replied:
ID review: Change the "using" to "with" or "by using"

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_UI_PATH_INVALID_CWWKO1675E.explanation=The OpenAPI UI Path cannot be set to a value that contains invalid characters.
+OPEN_API_UI_PATH_INVALID_CWWKO1675E.useraction=Ensure the OpenAPI Path contains only A-Za-z0-9, _, -, ., /.
+
+OPEN_API_DOC_PATH_INVALID_CWWKO1676E=CWWKO1676: The OpenAPI Document path is invalid or contains invalid characters. The URL must be constructed using Unicode alphanumeric characters A-Za-z0-9, the underscore (_), dash (-), forward slash (/) and period (.).
```
L2 suggested edit to remove the word "the" prior to "underscore", to maintain parallelism in the list of valid characters:
`OPEN_API_DOC_PATH_INVALID_CWWKO1676E=CWWKO1676: The OpenAPI Document path is invalid or contains invalid characters. The URL must be constructed using Unicode alphanumeric characters A-Za-z0-9, underscore (_), dash (-), forward slash (/), or period (.).`

helyarp replied:
ID review: Change the "using" to "with" or "by using"

posmith commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -98,3 +98,43 @@ OPENAPI_MERGE_INVALID_NAME_CWWKO1666W.useraction=Review the error text to identi
+OPEN_API_DOC_PATH_INVALID_CWWKO1676E.explanation=The OpenAPI Document Path cannot be set to a value that contains invalid characters.
+OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains only A-Za-z0-9, _, -, ., /.
+
+OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E=CWWKO1673E: path segments "/." or "/.." are invalid for OpenAPI endpoints.
+OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.explanation=OpenAPI path segments "/." or "/.." are invalid.
+OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.useraction=Remove any path "/." or "/.." segments from the path.
```
Use singular subject/verb in conjunction with "or".

L2 suggested edits:
```
OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E=CWWKO1673E: Path segment "/." or "/.." is invalid for OpenAPI endpoints.
OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.explanation=OpenAPI path segment "/." or "/.." is invalid.
OPEN_API_PATH_SEGMENT_INVALID_CWWKO1677E.useraction=Remove any path "/." or "/.." segment from the path.
```

posmith commented:
L2 message review completed

helyarp reviewed (commented):
Suggested changes

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,26 @@
+                   xmlns:ibmui="http://www.ibm.com/xmlns/appservers/osgi/metatype/ui/v1.0.0"
+                   localization="OSGI-INF/l10n/metatype">
+
+<OCD id="io.openliberty.microprofile.openapi" name="%mpOpenAPI" description="%mpOpenAPI.desc" ibm:alias="mpOpenAPI">
+    <metatype:AD id="docPath" name="%docPath" description="%docPath.desc" default="/openapi" required="false" type="String" />
+    <metatype:AD id="uiPath" name="%uiPath" description="%uiPath.desc" required="false" type="String" />
+</OCD>
```
These need the `ibm:beta` tag on them.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/bnd.bnd:
```diff
@@ -19,11 +19,34 @@ Bundle-SymbolicName: io.openliberty.microprofile.openapi.internal.common
+    io.openliberty.microprofile.openapi.internal.common;version="1.0"
+
+Include-Resource: \
+	OSGI-INF=resources/OSGI-INF
```
Fix indentation

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/WABConfigManager.java:
```diff
@@ -0,0 +1,101 @@
+        final Dictionary<String, String> props = new Hashtable<String, String>();
+        props.put(WABConfiguration.CONTEXT_NAME, contextName);
+        props.put(WABConfiguration.CONTEXT_PATH, contextPath);
+        if (wabConfigReg == null) {
+            wabConfigReg = context.registerService(WABConfiguration.class, new WABConfiguration() {
+            }, props);
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                registerEvent(props);
+            }
+        } else {
+            wabConfigReg.setProperties(props);
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                modifiedEvent(props);
+            }
+        }
```
This doesn't make sense. The way this class is written there's no opportinuty for `wabConfigReg` to be `null` at this point. The properties are effectively constant.

Azquelt replied:
Rewrote this so that the path can be changed.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,181 @@
+            Tr.info(tc, "OPEN_API_UI_PATH_UPDATE_CWWKO1668I", docPath);
+        }
+    }
+    /**
+     * Validate the OpenAPI path to ensure it is Valid, contains only valid characters and mets certain rules.
+     *
+     * @param path Path to be validated
+     * @param id   Which OpenAPI bundle we are validating the path for.
+     * @return
+     */
+    public static boolean validatePath(String path, EndpointId id) {
```
Update javadoc to indicate that this method also prints warnings.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UICustomPathTest.java:
```diff
@@ -0,0 +1,116 @@
+
+        // Check the headerbar colour
+        WebElement headerbar = waitForElement(driver, By.cssSelector("div.headerbar"));
+        assertEquals("Headerbar colour", "rgba(25, 28, 44, 1)", headerbar.getCssValue("background-color"));
```
Use `Color.fromString()` here as in `UIBasicTest`

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,181 @@
+     * @param path
+     * @return
+     */
+    public static String resolvePath(String path) {
```
This isn't really resolving. Suggest renaming to `normalizePath`.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManager.java:
```diff
@@ -0,0 +1,181 @@
+        docWabConfigManager.unregister();
+    }
+
+    private void resolvePathFromProperties(Map<String, Object> properties) {
```
This isn't really resolving. Suggest renaming to `getPathsFromProperties` and adding javadoc.

Azquelt commented on dev/io.openliberty.microprofile.openapi.internal.common/test/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManagerTest.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Test
+    public void validatePathFeaturePaths(){
+        assertTrue(OpenAPIEndpointManager.validatePath(METRICS_PATH,id1));
```
`validatePath` has been changed from a `String` argument to an `EndpointId` argument.

Almost every test in this class needs updated to compile.

Azquelt commented:
#libby
#build

Azquelt committed:
MP OpenAPI config tests

Add FAT tests for various configuration stages and steps for the new customization of OpenAPI endpoints

Tests invalid path characters - special characters and `.` based paths
Conflicting paths between Doc and UI and with other applications

Add in example log entries relating to the expected App startup failures

Azquelt commented:
#libby
#build

Azquelt commented:
@abutch3r Please review the changes I've made.

I don't think github will let you add an approval because it's your PR, but if you're happy with my changes then I'll add an approval.

abutch3r commented on dev/io.openliberty.microprofile.openapi.internal.common/test/io/openliberty/microprofile/openapi/internal/common/OpenAPIEndpointManagerTest.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Test
+    public void validatePathFeaturePaths() {
```
As the feature potential conflict check has been removed, this test doesn't serve any purpose any more as this was to check that we didn't accidentally block the setting of these values.

Azquelt committed:
Move MP OpenAPI servlets to separate bundles

We want to be able to change the context root of the MP OpenAPI servlet
at runtime. When we do this, the bundle serving the servlet gets
restarted.

This is fine for the servlets but not for the other components in the
same bundle.

Azquelt commented:
Added @dmuelle's suggested changes from #25511

Azquelt commented:
Rebased on `release`

#build
#libby

abutch3r commented:
#build

abutch3r commented:
#build

abutch3r committed:
Update returned index.html to use actual document endpoint

Add filter to allow for processing of response content

Update returned HTML to replace `/openapi` with the defined endpoint for the documents

Add no-store for the index.html page such that that page is fetched fresh every time

Update js such that it is no longer hardcoded to /openapi/ui for oauth2 html

Add ability to retrieve the actual endpoints that are being used

abutch3r committed:
Improve handling of stopped bundles for custom files

Potential for an update to occur during openapi-ui bundle shutdown due to way openapi customization applies its updates which leads to unexpected nullpointer as the bundle is in a STOPPED state

abutch3r commented:
#libby #build

abutch3r commented:
latest doc updates based on feedback in https://github.com/OpenLiberty/open-liberty/pull/25511

Azquelt commented on dev/com.ibm.ws.openapi.ui/swagger-ui/dist/index.html:
```diff
@@ -5,7 +5,7 @@
     <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32" />
     <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16" />
-    <script defer src="liberty-swagger-ui-bundle-f4a5a5cf1b006a74f119.js"></script><link href="liberty-swagger-ui-a44cef3790da8bb191bd.css" rel="stylesheet">
+    <script defer src="liberty-swagger-ui-bundle-f46b8f492bc9c6c9d41c.js"></script><link href="liberty-swagger-ui-a44cef3790da8bb191bd.css" rel="stylesheet">
```
It looks like you've correctly rebuilt the UI after making your change, but we will need to run the manual regression test before merging this (once we have a good personal build): https://github.com/OpenLiberty/openapi-ui-test-app

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/com/ibm/ws/microprofile/openapi/servlet/filter/HtmlResponseWrapper.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public ServletOutputStream getOutputStream() {
+        if (outputStream == null) {
+            outputStream = new ServletOutputStream() {
+                @Override
+                public void write(int b) throws IOException {
+                    baos.write(b);
+                }
+                @Override
+                public void flush() throws IOException {
+                    baos.flush();
+                }
+                @Override
+                public void close() throws IOException {
+                    baos.close();
+                }
+                @Override
+                public boolean isReady() {
+                    return false;
+                }
+                @Override
+                public void setWriteListener(WriteListener var1) {
+                }
+            };
+        }
+
+        return outputStream;
+    }
+
+    @Override
+    public PrintWriter getWriter() throws IOException {
+        if (writer == null) {
+            writer = new PrintWriter(new OutputStreamWriter(baos, getCharacterEncoding()));
+        }
+
+        return writer;
+    }
```
Although the use case for this filter is limited, it would be good to throw `IllegalStateException` if both methods are called.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/css/OpenAPIUIBundlesUpdater.java:
```diff
@@ -12,38 +12,24 @@
-import org.osgi.framework.FrameworkUtil;
-
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
 import com.ibm.websphere.ras.annotation.Trivial;
 import com.ibm.wsspi.kernel.service.utils.FileUtils;
-
 import io.openliberty.microprofile.openapi20.internal.utils.CloudUtils;
 import io.openliberty.microprofile.openapi20.internal.utils.Constants;
 import io.openliberty.microprofile.openapi20.internal.utils.LoggingUtils;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleException;
+import org.osgi.framework.FrameworkUtil;
+
+import java.io.*;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
```
Check the formatting here. It looks like your settings are different to the project settings.

abutch3r replied:
IDE trying to be helpful - have undone the changes and changed IDE settings

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/css/OpenAPIUIBundlesUpdater.java:
```diff
@@ -295,7 +284,12 @@ private static void writeStreamToZip(InputStream stream, ZipOutputStream zos) th
     private static boolean waitForBundlesToStart(Set<Bundle> openAPIUIBundles) {
         try {
-            new OpenAPIUIBundlesListener(openAPIUIBundles).await();
+            BundleContext bundleContext = FrameworkUtil.getBundle(OpenAPIUIBundlesUpdater.class).getBundleContext();
+            if(bundleContext != null){
+                new OpenAPIUIBundlesListener(openAPIUIBundles, bundleContext).await();
+            } else {
+                return false;
+            }
```
We might want a comment here to say why the bundleContext may be null and why it's safe to ignore it and return `false` without logging.

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/com/ibm/ws/microprofile/openapi/servlet/filter/OpenAPIEndpointFilter.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ *
+ */
+public class OpenAPIEndpointFilter implements Filter {
+
+    private static final TraceComponent tc = Tr.register(OpenAPIEndpointFilter.class);
```
We have two copies of this class. Should it go in the OpenAPI UI bundle instead so we'd only need one copy?

I think it would only need `OpenAPIEndpointProvider` from the common bundle.

Azquelt replied:
Ah, possibly we've done it this way around so that we don't need to worry about transforming the UI bundle to work with Jakarta. It does look like we already have another (identical) Filter in both projects.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/WEB-INF/web.xml:
```diff
@@ -26,5 +26,12 @@
     <url-pattern>/*</url-pattern>
   </filter-mapping>
-  
+  <filter>
+    <filter-name>openapi-ui-filter</filter-name>
+    <filter-class>com.ibm.ws.microprofile.openapi.servlet.filter.OpenAPIEndpointFilter</filter-class>
+  </filter>
```
I'm surprised this works, I'm not sure how this bundle can see the `OpenAPIEndpointFilter` class

abutch3r committed:
Updates based on feedback

Add in comments to explain decisions around handling of bundlecontexts

Add in IllegalStateExceptions to wrapper

Restructure the filters to handle the case were CSS Customization causes the underlying bundle to restart meaning that the filter is attempted to be used while its bundle is in a STOPPED state.

Undo Intellij import optimization

abutch3r commented:
#build

abutch3r commented:
#build

Azquelt commented:
#libby

Azquelt commented:
We need to run the [manual testing steps](https://github.com/OpenLiberty/openapi-ui-test-app) on the result of the personal build since the UI was rebuilt.

If everything looks good there then this is ready to merge after the next release is tagged.

abutch3r commented:
Manual Testing results using output of build

## Open Liberty UI Customizations

### MicroProfile UI
- [x] UI Loads on https://localhost:9443/openapi/ui/
- [x] Forced reload causes all resources to be reloaded
- [x] OpenLiberty logo is at the top of the page
- [x] Top of the page has the title and description for the application but no links
- [x] `OAS3` icon surrounding colour is `#5d8203` via element inspection
- [x] Footer is present
- [x] Footer contains links for `Terms of service`, `Test Author - Website`, `Contact Test Author` and `Example license`
- [x] No `Valid` button in footer
- Operation Block colours:
  - [x] POST background is `rgba(32,128,80,.1)` (div class = `opblock opblock-post`)
  - [x] POST border is #208050` (div class = `opblock opblock-post`)
  - [x] POST Icon background colour is `#208050` (div class = `opblock-summary-method`)
  - [x] PUT background is `rgba(177,99,3,.1)` (div class = `opblock  opblock-put`)
  - [x] PUT border colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] PUT Icon background colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] DELETE background is `rgba(224,7,7,.1)` (div class = `opblock  opblock-delete`)
  - [x] DELETE border colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] DELETE Icon background colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] GET background is `rgba(31,111,240,.1)` (div class = `opblock  opblock-get`)
  - [x] GET border colour is `#1f6ff0` (div class=`opblock-summary-method`)
  - [x] GET Icon background colour is `#1f6ff0` (div class=`opblock-summary-method`)

### Openapi 3.x UI
Public UI
- [x] UI loads https://localhost:9443/api/explorer/
- [x] Forced reload causes all resources to be reloaded
- [x] OpenLiberty logo is at the top of the page
- [x] The Filter box is present at the top of the page
- [x] Filter using `openapi-ui-test-app`  of prefixes of returns the operations
- [x] Filter using anything other value returns no operations
- [x] Top of the Page has `Liberty REST APIs`
- [x] `OAS3` icon surrounding colour is `#5d8203` via element inspection
- [x] No `Valid` button in page footer
- Operation Block colours:
  - [x] POST background is `rgba(32,128,80,.1)` (div class = `opblock opblock-post`)
  - [x] POST border is #208050` (div class = `opblock opblock-post`)
  - [x] POST Icon background colour is `#208050` (div class = `opblock-summary-method`)
  - [x] PUT background is `rgba(177,99,3,.1)` (div class = `opblock  opblock-put`)
  - [x] PUT border colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] PUT Icon background colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] DELETE background is `rgba(224,7,7,.1)` (div class = `opblock  opblock-delete`)
  - [x] DELETE border colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] DELETE Icon background colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] GET background is `rgba(31,111,240,.1)` (div class = `opblock  opblock-get`)
  - [x] GET border colour is `#1f6ff0` (div class=`opblock-summary-method`)
  - [x] GET Icon background colour is `#1f6ff0` (div class=`opblock-summary-method`)

Private UI
- [x] UI load on https://localhost:9443/ibm/api/explorer/ having provided credentials outlined in README
- [x] Forced reload causes all resources to be reloaded
- [x] OpenLiberty logo is at the top of the page
- [x] The Filter box is present at the top of the page
- [x] Filter using `openapi-ui-test-app`  of prefixes of returns the operations
- [x] Filter using anything other value returns no operations
- [x] Top of the Page has `Liberty REST APIs`
- [x] `OAS3` icon surrounding colour is `#5d8203` via element inspection
- [x] No `Valid` button in page footer
- [x] Operation Block colours:
  - [x] POST background is `rgba(32,128,80,.1)` (div class = `opblock opblock-post`)
  - [x] POST border is #208050` (div class = `opblock opblock-post`)
  - [x] POST Icon background colour is `#208050` (div class = `opblock-summary-method`)
  - [x] PUT background is `rgba(177,99,3,.1)` (div class = `opblock  opblock-put`)
  - [x] PUT border colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] PUT Icon background colour is `#b16303` (div class=`opblock-summary-method`)
  - [x] DELETE background is `rgba(224,7,7,.1)` (div class = `opblock  opblock-delete`)
  - [x] DELETE border colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] DELETE Icon background colour is `#e00707` (div class=`opblock-summary-method`)
  - [x] GET background is `rgba(31,111,240,.1)` (div class = `opblock  opblock-get`)
  - [x] GET border colour is `#1f6ff0` (div class=`opblock-summary-method`)
  - [x] GET Icon background colour is `#1f6ff0` (div class=`opblock-summary-method`)

## Functionality Testing

### MicroProfile UI
Access https://localhost:9443/openapi/ui
- [x] Two Server values in Server dropdown
- Select `HTTPS` server url
- Expand `GET /{id}` operation
  - [x] Two responses - `200` and `404` displayed
  - Change `media type` in drop down
    - [x] Example changes from either JSON -> XML or XML->JSON
  - [x]  Can switch between `Example Value` and `Schema`
  - [x] Attempt to provide an ID value and should be prohibited from doing so
  - Click `Try Out`
    -  Provide an ID value of `1`
    - Click `Execute` and get a 
      -  [x] `200` response
    - Change ID value to `0`
    - Click `Execute`
     - [x] `404` response
- Expand `POST /private`
  - [x] Padlock Icon is `Unlocked`
  - Click `Try Out`
    - Click execute 
    - [x] `401` response
-  Click `Authorize` and go to `oauth` section
  - Provide `Client_id` provided in README
  - Provide `Client_secret` provided in README
  - Select Test Scope
  - Click `authorize`
    - Login with user credentials provided in README
    - click `allow once`
    - [x] returned to the OpenAPI UI
  - Close modal
- [x] All `/private*` operations now have a locked padlock
- Expand `POST /private` 
  - Click `Try Out`
  - [x] Click `Execute` and get a `200` response
- [x] Expand `GET /apiKeyNeeded`
  - [x] Padlock is `unlocked`
  - [x] Description is `Get a random record`
  - [x] Click Try Out
    - [x] Click Execute and get a `403` response
- Click `Authorize`
  - Enter apikey from README
  - Click `Authorize`
  - Close modal

### OpenAPI 3.x UI
Access https://localhost:9443/api/explorer
- [x] Two Server values in Server dropdown
- Select `HTTPS` server url
- Expand `GET /{id}` operation
  - [x] Two responses - `200` and `404` displayed
  - Change `media type` in drop down
    - [x] Example changes from either JSON -> XML or XML->JSON
  - [x]  Can switch between `Example Value` and `Schema`
  - [x] Attempt to provide an ID value and should be prohibited from doing so
  - Click `Try Out`
    -  Provide an ID value of `1`
    - Click `Execute` and get a 
      -  [x] `200` response
    - Change ID value to `0`
    - Click `Execute`
     - [x] `404` response
- Expand `POST /private`
  - [x] Padlock Icon is `Unlocked`
  - Click `Try Out`
    - Click execute 
    - [x] `401` response
-  Click `Authorize` and go to `oauth` section
  - Provide `Client_id` provided in README
  - Provide `Client_secret` provided in README
  - Select Test Scope
  - Click `authorize`
    - Login with user credentials provided in README
    - click `allow once`
    - [x] returned to the OpenAPI UI
  - Close modal
- [x] All `/private*` operations now have a locked padlock
- Expand `POST /private` 
  - Click `Try Out`
  - [x] Click `Execute` and get a `200` response
- [x] Expand `GET /apiKeyNeeded`
  - [x] Padlock is `unlocked`
  - [x] Description is `Get a random record`
  - [x] Click Try Out
    - [x] Click Execute and get a `403` response
- Click `Authorize`
  - Enter apikey from README
  - Click `Authorize`
  - Close modal
- Expand `/apiKeyNeeded`
  - [x] Operation has a locked padlock
  - Click `Try Out`
    - Click `Execute`
    -  [x] `200` response

## Customer provided customization

MicroProfile UI
- If the server is running, stop the server
- Rename the folder ` example-customization` in `src/main/liberty/config` to `mpopenapi`
- Start the server
- Access https://localhost:9443/openapi/ui/ 
  - Perform a forced page refresh
  - [x] Header bar colour has changed
  - [x] Logo has changed 

OpenAPI-3.x UI
- If the server is running, stop the server
- Rename the folder ` example-customization` (or mpopenapi if renamed for MicroProfile testing) in `src/main/liberty/config` to `openapi-3.1` 
- Start the server
- Access https://localhost:9443/api/explorer 
  - Perform a forced page refresh
  - [x] Header bar colour has changed
  - [x] Logo has changed 

## Endpoint customization
(Microprofile only)

- Update mpOpenAPI config with new `docPath`
- Start server
- Access https://localhost:9443/<docpath>/ui
- [x] Able to see page
- [x] Page shows api operations
- [x] Default UI customizations are in place
- Update mpOpenAPI config with different `uiPath`
- Update Oauth Provider config redirect path with uiPath value
- Start server
- Access https://localhost:9443/<uiPath>
- [x] Able to see page
- [x] Page shows api operations
-  Click `Authorize` and go to `oauth` section
  - Provide `Client_id` provided in README
  - Provide `Client_secret` provided in README
  - Select Test Scope
  - Click `authorize`
    - Login with user credentials provided in README
    - click `allow once`
    - [x] returned to the OpenAPI UI
  - Close modal

abutch3r commented:
build Test failures unrelated to openapi and are typically related to issues loading arquillian operations and these passed in the previous PB and no change has occured which would now start them failing

abutch3r commented:
#libby
